### PR TITLE
port to FreeBSD armv6/7

### PR DESCRIPTION
**Pull request by MikaelUrankar**

### CODE REVIEW

File: make/lib/Lib-jdk.hotspot.agent.gmk
Comment (battleblow) [inline make/lib/Lib-jdk.hotspot.agent.gmk]: Is this a lack of porting the serviceability agent to arm32 or does it not work on arm32 in general (e.g. Linux also doesn't support it)

File: src/hotspot/cpu/arm/icache_arm.cpp
Comment (battleblow) [inline src/hotspot/cpu/arm/icache_arm.cpp]: I thought per comments on the previous PR that `__builtin___clear_cache` would work on arm32?

File: src/hotspot/cpu/arm/vm_version_arm_32.cpp
Comment (battleblow) [inline src/hotspot/cpu/arm/vm_version_arm_32.cpp:113]: Should this be `#ifdef KUSER_HELPER_VERSION_ADDR` instead?  I'd prefer this to be in terms of functionality rather than OS if that makes sense.

File: src/hotspot/os/bsd/os_perf_bsd.cpp
Comment (battleblow) [inline src/hotspot/os/bsd/os_perf_bsd.cpp]: Haven't we already included this two lines above?  Maybe this is a leftover from trying to avoid the frame definition conflict in other ways.

File: src/hotspot/os/bsd/os_perf_bsd.cpp
Comment (battleblow) [inline src/hotspot/os/bsd/os_perf_bsd.cpp]: This code should be machine independent, why are we avoiding it on arm32?

**Review by battleblow — COMMENTED** [review]
What I'd like to also see with this, now that we've resolved the frame ugliness, is either a volunteer to help maintain this or instructions on how I can at least test compile this going forward.  I currently test amd64 regularly and aarch64 and ppc64 every so often, but I don't have a way to test arm32 right now.

**Comment by battleblow** [comment]
> What I'd like to also see with this, now that we've resolved the frame ugliness, is either a volunteer to help maintain this or instructions on how I can at least test compile this going forward. I currently test amd64 regularly and aarch64 and ppc64 every so often, but I don't have a way to test arm32 right now.

Also, is someone prepared to forward port this, at least to the current repo?

File: src/hotspot/cpu/arm/icache_arm.cpp
Comment (MikaelUrankar) [reply src/hotspot/cpu/arm/icache_arm.cpp]: yes, sorry

File: make/lib/Lib-jdk.hotspot.agent.gmk
Comment (MikaelUrankar) [reply make/lib/Lib-jdk.hotspot.agent.gmk]: I don't remember, I'll try to build it again.

File: src/hotspot/cpu/arm/vm_version_arm_32.cpp
Comment (MikaelUrankar) [reply src/hotspot/cpu/arm/vm_version_arm_32.cpp:113]: KUSER_HELPER_VERSION_ADDR is defined for every platform, I can ifdef it in https://github.com/battleblow/openjdk-jdk11u/blob/bsd-port/src/hotspot/cpu/arm/vm_version_arm.hpp#L92 and `ifdef KUSER_HELPER_VERSION_ADDR` here. It's really linux specific

File: src/hotspot/os/bsd/os_perf_bsd.cpp
Comment (MikaelUrankar) [reply src/hotspot/os/bsd/os_perf_bsd.cpp]: some leftover, I'm struggling with git

File: src/hotspot/os/bsd/os_perf_bsd.cpp
Comment (MikaelUrankar) [reply src/hotspot/os/bsd/os_perf_bsd.cpp]: some leftover, I'm struggling with git

**Comment by MikaelUrankar** [comment]
> What I'd like to also see with this, now that we've resolved the frame ugliness, is either a volunteer to help maintain this or instructions on how I can at least test compile this going forward. I currently test amd64 regularly and aarch64 and ppc64 every so often, but I don't have a way to test arm32 right now.

I can maintain it.

> Also, is someone prepared to forward port this, at least to the current repo?

Do you mean porting to jdk12, jdk13 etc? I can do it.

File: src/hotspot/cpu/arm/vm_version_arm_32.cpp
Comment (battleblow) [reply src/hotspot/cpu/arm/vm_version_arm_32.cpp:113]: Makes sense.  Let's leave if as `#ifdef __linux__` then.

**Comment by battleblow** [comment]
> > What I'd like to also see with this, now that we've resolved the frame ugliness, is either a volunteer to help maintain this or instructions on how I can at least test compile this going forward. I currently test amd64 regularly and aarch64 and ppc64 every so often, but I don't have a way to test arm32 right now.
> 
> I can maintain it.

Thanks!  I will do a basic merge when there are changes to the Linux version.  That mostly works well, but there are times that doesn't work out perfectly.  So periodically test compiling it is great.

> > Also, is someone prepared to forward port this, at least to the current repo?
> 
> Do you mean porting to jdk12, jdk13 etc? I can do it.

I mostly mean to the jdk repo.  That is the current head of development and will at this point be what becomes jdk17.  I'll leave it up to you whether you want to target all the intermediate releases (jdk12, jdk13, etc.) as well, but I don't know that they are getting a lot of use.

File: make/lib/Lib-jdk.hotspot.agent.gmk
Comment (MikaelUrankar) [reply make/lib/Lib-jdk.hotspot.agent.gmk]: It doesn't seem to be implemented in linux, I can't find anything in  `src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/linux` or in `src/jdk.hotspot.agent/linux/native/libsaproc/LinuxDebuggerLocal.c`

File: make/autoconf/flags-cflags.m4
Comment (battleblow) [inline make/autoconf/flags-cflags.m4:742]: Are these flags that you'd want turned on for Linux if it was building using Clang on arm32?

File: make/lib/Lib-jdk.hotspot.agent.gmk
Comment (battleblow) [reply make/lib/Lib-jdk.hotspot.agent.gmk]: I downloaded the AdoptOpenJDK distribution for arm32 and it does contain libsaproc.so, so it seems like Linux is building the pieces for this even though it may not have arch specific files for it.  Does it error out building on FreeBSD?

**Review by battleblow — COMMENTED** [review]
Thanks for updating this Mikael.  My only real concern is the special casing of whether to build libsaproc or not.  I'd like to understand how that works on Linux better since it does not disable it like that and appears to build it.

File: make/autoconf/flags-cflags.m4
Comment (MikaelUrankar) [reply make/autoconf/flags-cflags.m4:742]: I don't know, there is the same thing for gcc just below: https://github.com/battleblow/openjdk-jdk11u/blob/bsd-port/make/autoconf/flags-cflags.m4#L751-L755
error without it:
```
ad_arm.cpp:19815:4: error: "ARM must be defined"
#  error "ARM must be defined"
   ^
1 error generated.
gmake[4]: *** [lib/CompileJvm.gmk:160: /usr/ports/java/openjdk11/work/openjdk-jdk11u-jdk-11.0.10-9-1/build/bsd-arm-normal-server-release/hotspot/variant-server/libjvm/objs/ad_arm.o] Error 1
gmake[4]: Leaving directory '/usr/ports/java/openjdk11/work/openjdk-jdk11u-jdk-11.0.10-9-1/make/hotspot'
```

the error for libsa is:
```
/usr/ports/java/openjdk11/work/openjdk-jdk11u-jdk-11.0.10-9-1/src/jdk.hotspot.agent/bsd/native/libsaproc/BsdDebuggerLocal.c:383:37: error: use of undeclared identifier 'NPRGREG'
  array = (*env)->NewLongArray(env, NPRGREG);
```

Comment (MikaelUrankar) [reply make/autoconf/flags-cflags.m4:742]: Ok, linux ifdef Java_sun_jvm_hotspot_debugger_linux_LinuxDebuggerLocal_getThreadIntegerRegisterSet0 for supported arch:
https://github.com/battleblow/openjdk-jdk11u/blob/bsd-port/src/jdk.hotspot.agent/linux/native/libsaproc/LinuxDebuggerLocal.c#L399

**Review by battleblow — APPROVED** [review]
Thanks Mikael, looks good to me!

If you're interested, it looks like the native support for aarch64 got recently backported to JDK8.  I haven't tried copying over the OS specific code and getting it working on FreeBSD yet.

Also, do you want me to push a release of JDK11 based on this now or wait for the next Oracle JDK11 release?

**Comment by MikaelUrankar** [comment]
Thanks a lot for the review. It can wait the next openjdk release.

**Pull request merged**
